astropy/io/ascii/core.py

```diff
@@ -1035,14 +1040,25 @@ def _validate_and_copy(col, converters):
         return converters_out
 
     def _convert_vals(self, cols):
-        for col in cols:
-            # If a specific dtype was specified for a column, then use that
-            # to set the defaults, otherwise use the generic defaults.
-            default_converters = ([convert_numpy(col.dtype)] if col.dtype
-                                  else self.default_converters)
-
-            # If the user supplied a specific convert then that takes precedence over defaults
-            converters = self.converters.get(col.name, default_converters)
+        # Loads of things happen before that change the number of columns.
+        # This is where we need the dtype that was passed in, so this is
+        # where we test.
+        if (self.dtype is not None) and len(self.dtype) != len(cols):
+            raise ValueError(f"Number of columns returned ({len(cols)}) does not match number of dtypes set ({len(self.dtype)}).")
+        for i, col in enumerate(cols):
+            # If the user set a dtype argument, that overrides everything else
+            if self.dtype is not None:
+                converters = [convert_numpy(self.dtype[i])]
```

Comment:
Are we sure we want to index the `dtype` by number rather than name?

Reply:
If we want to access by name, it needs to be a list of tuples or a dict, while the current implementation is meant to mirror the `Table(names=[...], dtype=[...])` API.

Reply:
Fair enough!